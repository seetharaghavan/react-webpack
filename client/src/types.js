export const  REQUEST_LOG_IN = "REQUEST_LOG_IN"; 
export const  LOG_IN = "LOG_IN"; 
export const  LOG_OUT = "LOG_OUT"; 
export const  LOG_IN_FAILED = "LOG_IN_FAILED"; 

export const  ADD_NOTE = "ADD_NOTE";
export const  EDIT_NOTE = "EDIT_NOTE"; 
export const  REMOVE_NOTE = "REMOVE_NOTE";

export const  GET_NOTES = "GET_NOTES"; 
export const  SELECT_NOTE = 'SELECT_NOTE'; 
export const  ERROR_NOTE = 'ERROR_NOTE'; 

export const  NOTE_REQUEST = "NOTE_REQUEST"; 

export const  SIDEBAR_TOGGLE = 'SIDEBAR_TOGGLE'; 

export const  NOTE_FORM_RESET = "NOTE_FORM_RESET"; 

export const  BASE_URL = "http://localhost:4000"; 