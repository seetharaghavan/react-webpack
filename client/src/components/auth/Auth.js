import React from 'react'; 
import { useDispatch, useSelector } from 'react-redux';
import { login } from '../../services/actions';
import { REQUEST_LOG_IN } from '../../types';

const Auth = () => {
  const [loginInfo, setLoginInfo] = React.useState({
    userName: "",
    password: "",
  });

  const [errors, setErrors] = React.useState({ field: null, message: "" });

  const authState = useSelector((state) => state.authState);

  const dispatch = useDispatch();

  const setFormValue = (event) => {
    const { name, value } = event.target;
    setLoginInfo((prev) => ({
      ...prev,
      [name]: value,
    }));
  };

  const onSubmit = (e) => {
    e.preventDefault();
    if (ValidateForm(loginInfo)) {
      dispatch({ type: REQUEST_LOG_IN });
      setTimeout(() => {
        dispatch(login(loginInfo));
      }, 1000);
    } else {
      return;
    }
  };

  const ValidateForm = (loginInfo) => {
    const { userName, password } = loginInfo;
    let isValid = false;
    if (userName.length === 0) {
      setErrors((prev) => ({
        ...prev,
        field: "userName",
        message: "Enter valid user name",
      }));
      return isValid;
    } else if (password.length === 0) {
      setErrors((prev) => ({
        ...prev,
        field: "password",
        message: "Enter valid password",
      }));
      return isValid;
    } else {
      isValid = true;
      setErrors((prev) => ({ ...prev, field: null, message: "" }));
    }
    return isValid;
  };

  return (
    <form onSubmit={onSubmit}>
      <div className="form-group">
        <label>Username *</label>
        <input
          type="text"
          className="form-control"
          placeholder="Enter Username"
          value={loginInfo.userName}
          name="userName"
          onChange={setFormValue}
          data-testid="userName"
          required
        />
        {errors.field === "userName" && (
          <small className="form-text text-danger">{errors.message}</small>
        )}
      </div>
      <div className="form-group">
        <label>Password *</label>
        <input
          type="password"
          className="form-control"
          placeholder="Enter Password"
          value={loginInfo.password}
          name="password"
          onChange={setFormValue}
          data-testid="password"
          required
        />
        {errors.field === "password" && (
          <small className="form-text text-danger">{errors.message}</small>
        )}

        {authState.err && (
          <small className="form-text text-danger">{authState.err}</small>
        )}
      </div>
      {authState.isLoading ? (
        <div date-testid="loading">Loading</div>
      ) : (
        <button type="submit" data-testid="auth-submit" className="btn btn-primary">
          Submit
        </button>
      )}
    </form>
  );
};

export default Auth; 
