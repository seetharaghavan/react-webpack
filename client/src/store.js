import {createStore, combineReducers, applyMiddleware} from 'redux'; 
import ReduxThunk  from 'redux-thunk'; 
import authReducer from './reducers/auth.reducer';
import noteReducer from './reducers/note.reducer';
import sideBarReducer from './reducers/sidebar.reducer';

const store = createStore( 
    combineReducers({
        authState: authReducer,
        noteState: noteReducer,
        sidebarState: sideBarReducer
    }),
    {},
    applyMiddleware(ReduxThunk)
)

export default store; 

