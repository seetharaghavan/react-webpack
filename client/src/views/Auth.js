import React from 'react'; 
import { useDispatch, useSelector } from 'react-redux';
import Auth from '../components/auth/Auth';
import { me } from '../services/actions';

const AuthView = ({history}) => {
  const authState = useSelector(state => state.authState); 
  const dispatch = useDispatch(); 
  React.useEffect(() => {
    if(authState.isLoggedIn){
      history.push('/notes'); 
    } else {
      let token = localStorage.getItem('token'); 
      if(token) dispatch(me())
    }
  }, [authState, dispatch])

  return (
    <div className="container mt-60" style={{flexDirection: 'column', alignItems: 'center', flex: 1, justifyContent: 'center'}}>
      <div className="row">
        <div className="col-md-4 md-offset-6">
          <h3>Login </h3>
          <Auth />
        </div>
      </div>
    </div>
  );
};

export default AuthView; 
