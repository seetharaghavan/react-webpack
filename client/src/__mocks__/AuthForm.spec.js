import React from "react";
import renderer from "react-test-renderer";
import localStorage from "./localstorage";
import { Provider } from "react-redux";
import store from "../store";
import Auth from "../components/auth/Auth";
import { render, fireEvent, waitFor, cleanup } from "@testing-library/react";
import { BASE_URL, LOG_IN } from "../types";
import axios from 'axios'; 
const {act} = renderer;


afterEach(cleanup);

const component = (
  <Provider store={store}>
    <Auth />
  </Provider>
);

const seedUser = {
  userName: "test",
  password: "test",
};

const originalError = global.console.error;
beforeAll(() => {
	global.console.error = jest.fn((...args) => {
		if (typeof args[0] === 'string' && args[0].includes('It looks like you\'re using the wrong act() around your test interactions')) {
	      return
	    }
	    return originalError.call(console, args)
	});
});

afterAll(() => {
  	global.console.error.mockRestore();
});

describe("Basic Ui", () => {
  //basic ui
  it("should render", async () => {
    const tree = await renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should have  3 input components", () => {
    const { getByTestId } = render(component);
    expect(getByTestId("auth-submit")).toHaveTextContent("Submit");
    expect(getByTestId("userName")).toBeRequired();
    expect(getByTestId("password")).toBeRequired();
  });
});

describe("Form behaviour", () => {
  it("invalid user name", async () => {
    const { getByTestId, getByText } = render(component);

    fireEvent.change(getByTestId("userName"), {
      target: { value: "" },
    });

    fireEvent.change(getByTestId("password"), {
      target: { value: "" },
    });
    fireEvent.click(getByTestId("auth-submit"));
    expect(getByText("Enter valid user name")).toBeInTheDocument();
  });

  it("invalid password", async () => {
    const { getByTestId, getByText } = render(component);

    fireEvent.change(getByTestId("userName"), {
      target: { value: seedUser.userName },
    });

    fireEvent.change(getByTestId("password"), {
      target: { value: "" },
    });

    fireEvent.click(getByTestId("auth-submit"));

    expect(getByText("Enter valid password")).toBeInTheDocument();
  });

  it("ideal  scenario", async () => {
    const { getByTestId, getByText, queryByText } = render(component);
    fireEvent.change(getByTestId("userName"), {
      target: { value: seedUser.userName },
    });

    fireEvent.change(getByTestId("password"), {
      target: { value: seedUser.password },
    });

    await act(async () => {
        fireEvent.click(getByTestId("auth-submit"));
    })

    
    expect(queryByText("Enter valid password")).not.toBeInTheDocument();
    expect(queryByText("Enter valid user name")).not.toBeInTheDocument();
    expect(getByText("Loading")).toBeInTheDocument();
    
    await waitFor(() => {
        store.dispatch({ type: LOG_IN, payload: {token: 'ABCDEFGHIJKLMNOPQRSTUVWXYZabc', userName: seedUser.userName}}); 
        expect(queryByText("Loading")).not.toBeInTheDocument()
    });
  });

});
