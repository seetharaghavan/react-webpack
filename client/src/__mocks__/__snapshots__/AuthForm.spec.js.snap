// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Ui should render 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label>
      Username *
    </label>
    <input
      className="form-control"
      data-testid="userName"
      name="userName"
      onChange={[Function]}
      placeholder="Enter Username"
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label>
      Password *
    </label>
    <input
      className="form-control"
      data-testid="password"
      name="password"
      onChange={[Function]}
      placeholder="Enter Password"
      required={true}
      type="password"
      value=""
    />
  </div>
  <button
    className="btn btn-primary"
    data-testid="auth-submit"
    type="submit"
  >
    Submit
  </button>
</form>
`;
